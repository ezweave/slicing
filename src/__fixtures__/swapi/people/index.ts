import { transformSwapiPeopleResponse } from '@slicing/transform/swapi';
import { People } from '@slicing/types';
import LarsOwenJSON from './lars-owen.json';

export const LarsOwen: People = transformSwapiPeopleResponse(LarsOwenJSON);
