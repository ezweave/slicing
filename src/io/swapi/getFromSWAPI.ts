import axios, { Axios, AxiosResponse } from 'axios';

export interface Getter {
  (path: string): (index: number) => Promise<AxiosResponse>
}

export const getFromSWAPI = (
  baseUrl: string = 'https://swapi.dev/api/',
  getFn: Axios['get'] = axios.get
): Getter => (
  path
) => async (
  index
) => getFn(`${baseUrl}${path}/${index}/`);
