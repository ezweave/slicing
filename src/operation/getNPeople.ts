import { flow, range } from 'lodash';
import { map as mapFP } from 'lodash/fp';

import { People } from '@slicing/types';

import { GetPeople, getPeople } from './getPeople';

export interface GetNPeople {
  (n: number, getPeopleFn?: GetPeople): Promise<People[] | unknown[]>;
}

export const getNPeople: GetNPeople = (
  numberOfPeople,
  getPeopleFn = getPeople,
) =>
  flow(
    (n) => range(0, n),
    mapFP((i) => getPeopleFn(i)),
    (promises: Promise<People>[]) => Promise.all(promises),
  )(numberOfPeople);
