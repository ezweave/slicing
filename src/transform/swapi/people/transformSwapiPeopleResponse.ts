import { transformAllFieldNamesToCamel } from 'field-caser';

import { People } from '@slicing/types';

export const transformSwapiPeopleResponse = (response: any) =>
  <People>transformAllFieldNamesToCamel(response);
