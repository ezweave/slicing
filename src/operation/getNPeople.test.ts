import { LarsOwen } from '@slicing/__fixtures__/swapi/people';
import { getNPeople } from './getNPeople';

describe(getNPeople, () => {
  it('returns n people', async () => {
    const mockGetPeople = jest.fn().mockResolvedValue(LarsOwen);
    const n = 10;
    const people = await getNPeople(n, mockGetPeople);
    expect(people.length).toEqual(n);
    expect(mockGetPeople).toHaveBeenCalledTimes(n);
    expect(people).toMatchSnapshot();
  });
});
