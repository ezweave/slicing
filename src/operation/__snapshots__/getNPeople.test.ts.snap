// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getNPeople returns n people 1`] = `
Array [
  Object {
    "birthYear": "52BBY",
    "created": "2014-12-10T15:52:14.024000Z",
    "edited": "2014-12-20T21:17:50.317000Z",
    "eyeColor": "blue",
    "films": Array [
      "https://swapi.dev/api/films/1/",
      "https://swapi.dev/api/films/5/",
      "https://swapi.dev/api/films/6/",
    ],
    "gender": "male",
    "hairColor": "brown, grey",
    "height": "178",
    "homeworld": "https://swapi.dev/api/planets/1/",
    "mass": "120",
    "name": "Owen Lars",
    "skinColor": "light",
    "species": Array [],
    "starships": Array [],
    "url": "https://swapi.dev/api/people/6/",
    "vehicles": Array [],
  },
  Object {
    "birthYear": "52BBY",
    "created": "2014-12-10T15:52:14.024000Z",
    "edited": "2014-12-20T21:17:50.317000Z",
    "eyeColor": "blue",
    "films": Array [
      "https://swapi.dev/api/films/1/",
      "https://swapi.dev/api/films/5/",
      "https://swapi.dev/api/films/6/",
    ],
    "gender": "male",
    "hairColor": "brown, grey",
    "height": "178",
    "homeworld": "https://swapi.dev/api/planets/1/",
    "mass": "120",
    "name": "Owen Lars",
    "skinColor": "light",
    "species": Array [],
    "starships": Array [],
    "url": "https://swapi.dev/api/people/6/",
    "vehicles": Array [],
  },
  Object {
    "birthYear": "52BBY",
    "created": "2014-12-10T15:52:14.024000Z",
    "edited": "2014-12-20T21:17:50.317000Z",
    "eyeColor": "blue",
    "films": Array [
      "https://swapi.dev/api/films/1/",
      "https://swapi.dev/api/films/5/",
      "https://swapi.dev/api/films/6/",
    ],
    "gender": "male",
    "hairColor": "brown, grey",
    "height": "178",
    "homeworld": "https://swapi.dev/api/planets/1/",
    "mass": "120",
    "name": "Owen Lars",
    "skinColor": "light",
    "species": Array [],
    "starships": Array [],
    "url": "https://swapi.dev/api/people/6/",
    "vehicles": Array [],
  },
  Object {
    "birthYear": "52BBY",
    "created": "2014-12-10T15:52:14.024000Z",
    "edited": "2014-12-20T21:17:50.317000Z",
    "eyeColor": "blue",
    "films": Array [
      "https://swapi.dev/api/films/1/",
      "https://swapi.dev/api/films/5/",
      "https://swapi.dev/api/films/6/",
    ],
    "gender": "male",
    "hairColor": "brown, grey",
    "height": "178",
    "homeworld": "https://swapi.dev/api/planets/1/",
    "mass": "120",
    "name": "Owen Lars",
    "skinColor": "light",
    "species": Array [],
    "starships": Array [],
    "url": "https://swapi.dev/api/people/6/",
    "vehicles": Array [],
  },
  Object {
    "birthYear": "52BBY",
    "created": "2014-12-10T15:52:14.024000Z",
    "edited": "2014-12-20T21:17:50.317000Z",
    "eyeColor": "blue",
    "films": Array [
      "https://swapi.dev/api/films/1/",
      "https://swapi.dev/api/films/5/",
      "https://swapi.dev/api/films/6/",
    ],
    "gender": "male",
    "hairColor": "brown, grey",
    "height": "178",
    "homeworld": "https://swapi.dev/api/planets/1/",
    "mass": "120",
    "name": "Owen Lars",
    "skinColor": "light",
    "species": Array [],
    "starships": Array [],
    "url": "https://swapi.dev/api/people/6/",
    "vehicles": Array [],
  },
  Object {
    "birthYear": "52BBY",
    "created": "2014-12-10T15:52:14.024000Z",
    "edited": "2014-12-20T21:17:50.317000Z",
    "eyeColor": "blue",
    "films": Array [
      "https://swapi.dev/api/films/1/",
      "https://swapi.dev/api/films/5/",
      "https://swapi.dev/api/films/6/",
    ],
    "gender": "male",
    "hairColor": "brown, grey",
    "height": "178",
    "homeworld": "https://swapi.dev/api/planets/1/",
    "mass": "120",
    "name": "Owen Lars",
    "skinColor": "light",
    "species": Array [],
    "starships": Array [],
    "url": "https://swapi.dev/api/people/6/",
    "vehicles": Array [],
  },
  Object {
    "birthYear": "52BBY",
    "created": "2014-12-10T15:52:14.024000Z",
    "edited": "2014-12-20T21:17:50.317000Z",
    "eyeColor": "blue",
    "films": Array [
      "https://swapi.dev/api/films/1/",
      "https://swapi.dev/api/films/5/",
      "https://swapi.dev/api/films/6/",
    ],
    "gender": "male",
    "hairColor": "brown, grey",
    "height": "178",
    "homeworld": "https://swapi.dev/api/planets/1/",
    "mass": "120",
    "name": "Owen Lars",
    "skinColor": "light",
    "species": Array [],
    "starships": Array [],
    "url": "https://swapi.dev/api/people/6/",
    "vehicles": Array [],
  },
  Object {
    "birthYear": "52BBY",
    "created": "2014-12-10T15:52:14.024000Z",
    "edited": "2014-12-20T21:17:50.317000Z",
    "eyeColor": "blue",
    "films": Array [
      "https://swapi.dev/api/films/1/",
      "https://swapi.dev/api/films/5/",
      "https://swapi.dev/api/films/6/",
    ],
    "gender": "male",
    "hairColor": "brown, grey",
    "height": "178",
    "homeworld": "https://swapi.dev/api/planets/1/",
    "mass": "120",
    "name": "Owen Lars",
    "skinColor": "light",
    "species": Array [],
    "starships": Array [],
    "url": "https://swapi.dev/api/people/6/",
    "vehicles": Array [],
  },
  Object {
    "birthYear": "52BBY",
    "created": "2014-12-10T15:52:14.024000Z",
    "edited": "2014-12-20T21:17:50.317000Z",
    "eyeColor": "blue",
    "films": Array [
      "https://swapi.dev/api/films/1/",
      "https://swapi.dev/api/films/5/",
      "https://swapi.dev/api/films/6/",
    ],
    "gender": "male",
    "hairColor": "brown, grey",
    "height": "178",
    "homeworld": "https://swapi.dev/api/planets/1/",
    "mass": "120",
    "name": "Owen Lars",
    "skinColor": "light",
    "species": Array [],
    "starships": Array [],
    "url": "https://swapi.dev/api/people/6/",
    "vehicles": Array [],
  },
  Object {
    "birthYear": "52BBY",
    "created": "2014-12-10T15:52:14.024000Z",
    "edited": "2014-12-20T21:17:50.317000Z",
    "eyeColor": "blue",
    "films": Array [
      "https://swapi.dev/api/films/1/",
      "https://swapi.dev/api/films/5/",
      "https://swapi.dev/api/films/6/",
    ],
    "gender": "male",
    "hairColor": "brown, grey",
    "height": "178",
    "homeworld": "https://swapi.dev/api/planets/1/",
    "mass": "120",
    "name": "Owen Lars",
    "skinColor": "light",
    "species": Array [],
    "starships": Array [],
    "url": "https://swapi.dev/api/people/6/",
    "vehicles": Array [],
  },
]
`;
